sn_node/src/quote.rs

```diff
@@ -37,36 +39,97 @@ impl Node {
+        return Err(Error::InvalidQuoteContent);
+    }
 
-        // check if the quote has expired
-        if quote.has_expired() {
-            return Err(Error::QuoteExpired(address.clone()));
-        }
+    // check if the quote has expired
+    if quote.has_expired() {
+        return Err(Error::QuoteExpired(address.clone()));
+    }
 
-        // check sig
-        let bytes = PaymentQuote::bytes_for_signing(
-            quote.content,
-            quote.cost,
-            quote.timestamp,
-            &quote.quoting_metrics,
-        );
-        let signature = quote.signature;
-        if !self.network.verify(&bytes, &signature) {
-            return Err(Error::InvalidQuoteSignature);
-        }
+    // check sig
+    let bytes = PaymentQuote::bytes_for_signing(
+        quote.content,
+        quote.cost,
+        quote.timestamp,
+        &quote.quoting_metrics,
+    );
+    let signature = quote.signature;
+    if !network.verify(&bytes, &signature) {
+        return Err(Error::InvalidQuoteSignature);
+    }
+
+    Ok(())
+}
+
+// Following metrics will be considered as client issue instead of node's bad quote.
+//   1, quote is not regarding the same chunk as ours
+//   2, quote is not around the same time as ours
+//   3, quote is no longer valid
+//
+// Following metrics will be considered as node's bad quote.
+//   1, Price calculation is incorrect
+//   2, QuoteMetrics doesn't match the historical quotes collected by self
+pub(crate) async fn quotes_verification(network: &Network, quotes: Vec<(PeerId, PaymentQuote)>) {
+    // Do nothing if self is not one of the quoters.
+    if let Some((_, self_quote)) = quotes
+        .iter()
+        .find(|(peer_id, _quote)| *peer_id == *network.peer_id)
+    {
+        let target_address =
+            NetworkAddress::from_chunk_address(ChunkAddress::new(self_quote.content));
+        if verify_quote_for_storecost(network, self_quote.clone(), &target_address).is_ok() {
+            let mut quotes_for_nodes_duty: Vec<_> = quotes
+                .iter()
+                .filter(|(peer_id, quote)| {
+                    let is_same_target = quote.content == self_quote.content;
+                    let is_not_self = *peer_id != *network.peer_id;
+                    let is_not_zero_quote = quote.cost != NanoTokens::zero();
+
+                    let time_gap = Duration::from_secs(10);
+                    let is_around_same_time = if quote.timestamp > self_quote.timestamp {
+                        self_quote.timestamp + time_gap > quote.timestamp
+                    } else {
+                        quote.timestamp + time_gap > self_quote.timestamp
+                    };
+
+                    is_same_target && is_not_self && is_not_zero_quote && is_around_same_time
+                })
+                .cloned()
+                .collect();
 
-        Ok(())
+            quotes_for_nodes_duty.retain(|(peer_id, quote)| {
+                let cost = calculate_cost_for_records(
```

Comment:
Could this not just be passed the full quote to simplify things here?

Reply:
yeah, I had the same feeling as well.
planned to raise a separate PR to refactor the API of `calculate_cost_for_records` as a whole,
so that can easily pass `quote` down.